### PR TITLE
Kafka codec: response encoder and decoder

### PR DESCRIPTION
**Pull request by adamkotwasinski**
Description: Kafka response codec
Part of https://github.com/envoyproxy/envoy/issues/2852
Based on existing request codec, including making common tooling more generic.

Risk Level: Low (Kafka code is unused right now)
Testing: automated tests, manual testing with real Kafka broker and client (dummy filter that deserializes the buffer, clears it, serializes the parse results into buffer again (what should be equivalent to what was received), and passes the buffer's contents further on - done for both requests & responses)
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by adamkotwasinski** [comment]
/wait the general response codec works, but there's some extra work I'd like to do regarding repeated code (and missing tests)

**Comment by adamkotwasinski** [comment]
Response codec.
Very similar to request codec, as we still keep the `onData` loop processing with a chain of `Parser` objects that return the next parser capable of processing payload or sentinel parser if we are dealing with garbage.
Generator script did not need any major changes apart from being made generic, the structures describing the kafka payloads are the same on both request & response sides.

Impl note:
It can be noticed that `RequestCodec::doParse` & `ResponseCodec::doParse` are very similar.
Same with `SentinelParser` / `SentinelResponseParser`.
I'd rather make this cleanup in a separate PR, as otherwise it introduces really a lot of noise.

File: source/extensions/filters/network/kafka/protocol_code_generator/request_codec_request_test_cc.j2 [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include "extensions/filters/network/kafka/external/requests.h"
 #include "extensions/filters/network/kafka/request_codec.h"
 
+#include "test/extensions/filters/network/kafka/buffer_based_test.h"
```
Comment (mattklein123) [inline source/extensions/filters/network/kafka/protocol_code_generator/request_codec_request_test_cc.j2]: Can you move this to the tests folder? Same below? I didn't catch this before but all test code should be in the test folder.

File: source/extensions/filters/network/kafka/response_codec.cc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#include "common/common/stack_array.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+void ResponseInitialParserFactory::expectResponse(const int16_t api_key,
+                                                  const int16_t api_version) {
+  std::lock_guard<std::mutex> lock{expected_responses_mutex_};
```
Comment (mattklein123) [inline source/extensions/filters/network/kafka/response_codec.cc]: Why do we need locks in this class?

Comment (adamkotwasinski) [reply source/extensions/filters/network/kafka/response_codec.cc]: I was unsure how many threads are operating on a filter, so I just took the safe approach :)
A filter needs to have a request codec to insert expected values into response codec (thru this call).
Is it the same thread that handles Filter's `onWrite` and `onData` calls? If yes, then I can remove that lock.

Comment (mattklein123) [reply source/extensions/filters/network/kafka/response_codec.cc]: Yeah everything is siloed in Envoy, so no locks should be needed.

**Comment by adamkotwasinski** [comment]
Fixes / merges applied.
There will be future PRs with refactoring only:
a) c++ code to address similarities between request/response parser/codec code,
b) python code to put source files for tests in `tests` directory.

**Comment by mattklein123** [comment]
Can you merge master? I think you might run into some clang-tidy issues around typedef vs. using. Should be simple fixes.

/wait

**Comment by adamkotwasinski** [comment]
Done!

**Pull request merged**